### PR TITLE
Add tests for sycl::link

### PR DESCRIPTION
**Pull request by vasilytric**
These tests do these verifications:
- if linkage is possible all kernels without optional features should be represented in the result bundle
- if linkage is possible only compatible kernels should be represented in the result bundle
- kernels, that provided to the kernel bundle should be invokable (if such kernel compatible with current device)
- exception was thrown in case device not from provided kernel bundle devices
- exception was thrown for zero devices passed
- exception was thrown for zero devices and zero bundles passed
- associated devices are equal to devices from obtained kernel bundle
- `property_list` existence and empty `property_list` can be passed or omitted

______________________________________
This PR depend on these PRs:
- https://github.com/KhronosGroup/SYCL-CTS/pull/218
- https://github.com/KhronosGroup/SYCL-CTS/pull/220
- https://github.com/KhronosGroup/SYCL-CTS/pull/226

### CODE REVIEW

**Review by ProGTX — APPROVED** [review]

File: tests/kernel_bundle/sycl_link_different_device_for_sycl_link.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  void get_info(test_base::info &out) const override {
+    set_test_info(out, TOSTRING(TEST_NAME), TEST_FILE);
+  }
+
+  /** execute the test
+   */
+  void run(util::logger &log) override {
+    const std::vector<sycl::device> devices{sycl::device::get_devices()};
+    if (devices.size() <= 1) {
+      log.note("Test skipped due to one device was found");
```
Comment (keryell) [inline tests/kernel_bundle/sycl_link_different_device_for_sycl_link.cpp:39]: Cf previous PR on this very same note message

File: tests/kernel_bundle/sycl_link_zero_device_and_zero_bundles.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    set_test_info(out, TOSTRING(TEST_NAME), TEST_FILE);
+  }
+
+  /** execute the test
+   */
+  void run(util::logger &log) override {
+    auto q = util::get_cts_object::queue();
+    vector_with_object_bundles zero_bundles{};
+
+    const std::vector<sycl::device> zero_device{};
```
Comment (keryell) [inline tests/kernel_bundle/sycl_link_zero_device_and_zero_bundles.cpp:39]:
```suggestion
    vector_with_object_bundles zero_bundles;

    const std::vector<sycl::device> zero_device;
```

File: tests/kernel_bundle/sycl_link_zero_devices.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    auto q = util::get_cts_object::queue();
+
+    const auto first_simple_kernel_id =
+        sycl::get_kernel_id<first_simple_kernel>();
+
+    auto kernel_bundle = sycl::get_kernel_bundle<sycl::bundle_state::object>(
+        q.get_context(), {first_simple_kernel_id});
+    vector_with_object_bundles vector_with_kb{kernel_bundle};
+
+    std::vector<sycl::device> zero_devices{};
```
Comment (keryell) [inline tests/kernel_bundle/sycl_link_zero_devices.cpp:46]:
```suggestion
    std::vector<sycl::device> zero_devices;
```

**Review by keryell — CHANGES_REQUESTED** [review]
That looks good.

File: tests/kernel_bundle/sycl_link_zero_devices.cpp (hunk 3 above)
Comment (vasilytric) [reply tests/kernel_bundle/sycl_link_zero_devices.cpp]: I've applied this comment in this commit: https://github.com/KhronosGroup/SYCL-CTS/pull/227/commits/050ffa8abb11c4f5b529f960904b20337a8b10e1

File: tests/kernel_bundle/sycl_link_zero_device_and_zero_bundles.cpp (hunk 2 above)
Comment (vasilytric) [reply tests/kernel_bundle/sycl_link_zero_device_and_zero_bundles.cpp]: I've applied this comment in this commit: https://github.com/KhronosGroup/SYCL-CTS/pull/227/commits/050ffa8abb11c4f5b529f960904b20337a8b10e1

File: tests/kernel_bundle/sycl_link_kernel_bundles_with_diff_ctx.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  void get_info(test_base::info &out) const override {
+    set_test_info(out, TOSTRING(TEST_NAME), TEST_FILE);
+  }
+
+  /** execute the test
+   */
+  void run(util::logger &log) override {
+    const std::vector<sycl::device> devices{sycl::device::get_devices()};
+    if (devices.size() <= 1) {
+      log.note("Test skipped due to one device was found");
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/sycl_link_kernel_bundles_with_diff_ctx.cpp]: I've applied this comment in this commit: https://github.com/KhronosGroup/SYCL-CTS/pull/227/commits/050ffa8abb11c4f5b529f960904b20337a8b10e1

File: tests/kernel_bundle/sycl_link_different_device_for_sycl_link.cpp (hunk 1 above)
Comment (vasilytric) [reply tests/kernel_bundle/sycl_link_different_device_for_sycl_link.cpp]: I've applied this comment in this commit: https://github.com/KhronosGroup/SYCL-CTS/pull/227/commits/050ffa8abb11c4f5b529f960904b20337a8b10e1

**Review by keryell — APPROVED** [review]
Thanks!

**Comment by vasilytric** [comment]
I am going to merge `SYCL-2020` branch into this branch after PR#226 will be merged because of the fact that this PR is depend on PR#226

**Pull request merged**